ndk/src/bitmap.rs

```diff
@@ -57,26 +59,33 @@ fn construct<T>(with_ptr: impl FnOnce(*mut T) -> i32) -> Result<T> {
     BitmapError::from_status(status).map(|()| unsafe { result.assume_init() })
 }
 
-#[repr(u32)]
-#[derive(Clone, Copy, Debug, PartialEq, Eq, IntoPrimitive, TryFromPrimitive)]
+#[repr(i32)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, IntoPrimitive, FromPrimitive)]
 #[allow(non_camel_case_types)]
 #[doc(alias = "AndroidBitmapFormat")]
+#[non_exhaustive]
 pub enum BitmapFormat {
     #[doc(alias = "ANDROID_BITMAP_FORMAT_NONE")]
-    NONE = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_NONE.0,
+    NONE = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_NONE.0 as i32,
     #[doc(alias = "ANDROID_BITMAP_FORMAT_RGBA_8888")]
-    RGBA_8888 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_8888.0,
+    RGBA_8888 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_8888.0 as i32,
     #[doc(alias = "ANDROID_BITMAP_FORMAT_RGB_565")]
-    RGB_565 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGB_565.0,
+    RGB_565 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGB_565.0 as i32,
     #[deprecated = "Deprecated in API level 13. Because of the poor quality of this configuration, it is advised to use ARGB_8888 instead."]
     #[doc(alias = "ANDROID_BITMAP_FORMAT_RGBA_4444")]
-    RGBA_4444 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_4444.0,
+    RGBA_4444 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_4444.0 as i32,
     #[doc(alias = "ANDROID_BITMAP_FORMAT_A_8")]
-    A_8 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_A_8.0,
+    A_8 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_A_8.0 as i32,
     #[doc(alias = "ANDROID_BITMAP_FORMAT_RGBA_F16")]
-    RGBA_F16 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_F16.0,
+    RGBA_F16 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_F16.0 as i32,
     #[doc(alias = "ANDROID_BITMAP_FORMAT_RGBA_1010102")]
-    RGBA_1010102 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_1010102.0,
+    RGBA_1010102 = ffi::AndroidBitmapFormat::ANDROID_BITMAP_FORMAT_RGBA_1010102.0 as i32,
+
+    // Use the SUCCESS discriminant, as no-one will be able to call `as i32` and only has access to
+    // the constants via `From` provided by `IntoPrimitive` which reads the contained value.
```

Comment:
If I recall correctly, the exact reason with one of the enums was that it went like:

```
enum XXX {
   Ok = 0,
   Err1 = -1,
   Err2 = -2,

   __Unknown(..) = (Err2 + 1),
```

This is done implicitly by the compiler, but would fail to compile because of overlap with `Err1`.

Reply:
It is `BitmapError`:

```rust
error[E0081]: discriminant value `-1` assigned more than once
  --> ndk/src/bitmap.rs:22:1
   |
22 | pub enum BitmapError {
   | ^^^^^^^^^^^^^^^^^^^^
...
26 |     BadParameter = ffi::ANDROID_BITMAP_RESULT_BAD_PARAMETER,
   |                    ---------------------------------------- `-1` assigned here
27 |     #[doc(alias = "ANDROID_BITMAP_RESULT_JNI_EXCEPTION")]
28 |     JniException = ffi::ANDROID_BITMAP_RESULT_JNI_EXCEPTION,
   |     ------------ discriminant for `__Unknown` incremented from this startpoint (`JniException` + 1 variant later => `__Unknown` = -1)
...
34 |     __Unknown(i32),
   |     --------- `-1` assigned here
```

Reply:
I'll clear out the rest though, should only be necessary in a few cases.

Reply:
Okey, yep, makes sense

Reply:
It doesn't really matter though _what_ discriminant we use, `.into()` from `IntoPrimitive` takes care of never returning it and the user is disallowed from using `as <primitive>`.

Reply:
By the way, for `aaudio_result_t` the variants are listed from lowest to highest, (since they have `base=-900` and then use `disconnected=base+1, illegal_argument=base+2` andsoforth). We could skip the discriminant here and have the Rust code generate one that might overlap with an upcoming variant in the NDK, again because `num_enum` should be able to deal with that and the tuple variant prevents calling a plain `as i32`.

```rust
pub const AAUDIO_ERROR_BASE: _bindgen_ty_50 = -900;
pub const AAUDIO_ERROR_DISCONNECTED: _bindgen_ty_50 = -899;
pub const AAUDIO_ERROR_ILLEGAL_ARGUMENT: _bindgen_ty_50 = -898;
pub const AAUDIO_ERROR_INTERNAL: _bindgen_ty_50 = -896;
pub const AAUDIO_ERROR_INVALID_STATE: _bindgen_ty_50 = -895;
pub const AAUDIO_ERROR_INVALID_HANDLE: _bindgen_ty_50 = -892;
pub const AAUDIO_ERROR_UNIMPLEMENTED: _bindgen_ty_50 = -890;
pub const AAUDIO_ERROR_UNAVAILABLE: _bindgen_ty_50 = -889;
pub const AAUDIO_ERROR_NO_FREE_HANDLES: _bindgen_ty_50 = -888;
pub const AAUDIO_ERROR_NO_MEMORY: _bindgen_ty_50 = -887;
pub const AAUDIO_ERROR_NULL: _bindgen_ty_50 = -886;
pub const AAUDIO_ERROR_TIMEOUT: _bindgen_ty_50 = -885;
pub const AAUDIO_ERROR_WOULD_BLOCK: _bindgen_ty_50 = -884;
pub const AAUDIO_ERROR_INVALID_FORMAT: _bindgen_ty_50 = -883;
pub const AAUDIO_ERROR_OUT_OF_RANGE: _bindgen_ty_50 = -882;
pub const AAUDIO_ERROR_NO_SERVICE: _bindgen_ty_50 = -881;
pub const AAUDIO_ERROR_INVALID_RATE: _bindgen_ty_50 = -880;
```

Our `__Unknown` variant implicitly gets discriminant `-879`.